<h1>Clustering Neighborhoods In Toronto -1
<h2> Coursera Capstone Project 
    <h3> Shiavm Wankhade

<h2> module1: Scraping Wikipedia page to obtain data and transform data into Dataframe

In [4]:
from bs4 import BeautifulSoup as bs
from urllib.request import urlopen as ureq
import requests
import lxml
import pandas as pd
from pandas import DataFrame
import numpy as np

In [5]:
my_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
r=requests.get(my_url)

<h2>Scraping the web file using BeautifulSoup package

In [6]:
page=bs(r.text,"html.parser")
page

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgMonthNamesShort":["","Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"wgRequestId":"XlFKiQpAMNIAAZcB4KkAAABD","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":935851093,"wgRevisionId":935851093,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontari

<h2>Let's find out Table in 'Html Page' indicated by 'Table Class'

In [7]:
rtable=page.table
rtable

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Harbourfront</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Lawrence_Heights" title="Lawrence Heights">Lawrence Heights</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North

In [8]:
results=rtable.find_all('tr')
nrows=len(results)
nrows

288

In [9]:
results[0:5]

[<tr>
 <th>Postcode</th>
 <th>Borough</th>
 <th>Neighbourhood
 </th></tr>, <tr>
 <td>M1A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M2A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M3A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td></tr>, <tr>
 <td>M4A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td></tr>]

<h2> Header of the dataframe is the first row of data(i.e. index[0])

In [10]:
header=results[0].text.split()
header

['Postcode', 'Borough', 'Neighbourhood']

<h2>Let's Check some rows in order to prepare data and extract cell into a dataframe

In [11]:
results[80].text

'\nM3J\nNorth York\nNorthwood Park\n'

In [12]:
results[80].text.split('\n')

['', 'M3J', 'North York', 'Northwood Park', '']

In [13]:
Postcode=results[80].text.split('\n')[1]
Postcode

'M3J'

In [14]:
Borough=results[80].text.split('\n')[2]
Borough

'North York'

In [15]:
Neighborhood=results[80].text.split('\n')[3]
Neighborhood

'Northwood Park'

<h2> Loop to extract all cells into dataframe

In [16]:
records=[]
n=1
while n < nrows:
    Postcode=results[n].text.split('\n')[1]
    Borough=results[n].text.split('\n')[2]
    Neighborhood=results[n].text.split('\n')[3]
    records.append((Postcode,Borough,Neighborhood))
    n=n+1
    
df=pd.DataFrame(records,columns=['Postcode','Borough','Neighborhood'])
df.head()
    

Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [17]:
df.shape

(287, 3)

In [18]:
df[df['Borough']=='Not assigned'].count()

Postcode        77
Borough         77
Neighborhood    77
dtype: int64

In [19]:
# drops those rows where 'Not assigned' appears in column '[Borough]'
df1=df[~df.Borough.str.contains("Not assigned")]
df1=df1.reset_index(drop=True)

In [20]:
df1.head(5)

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M6A        North York  Lawrence Heights
4      M6A        North York    Lawrence Manor

In [21]:
df1.tail(5)

Postcode    Borough              Neighborhood
205      M8Z  Etobicoke  Kingsway Park South West
206      M8Z  Etobicoke                 Mimico NW
207      M8Z  Etobicoke        The Queensway West
208      M8Z  Etobicoke     Royal York South West
209      M8Z  Etobicoke            South of Bloor

In [22]:
df1.shape

(210, 3)

<h2> Now lets replace "Not assigned" Neighborhood Same as in 'Borough'

In [23]:
df1.loc[df1['Neighborhood'] == 'Not assigned', 'Neighborhood'] = df1['Borough']


In [24]:
df1.head()

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M6A        North York  Lawrence Heights
4      M6A        North York    Lawrence Manor

<h2>Now , we will consolidate the dataframe to the unique Postal Codes.
<h3>As per instructions, we will look for repeated postalcodes and consolidate into one postalcode with the name of the neighbourhoods merged and separated by comma.
<h3>Let's examine de dataframe first to set up the best loop algorithm. How many unique postalcodes , Boroughs and Neighbourhood names are there?

In [25]:
postcodes = df1['Postcode'].nunique()
boroughs = df1['Borough'].nunique()
neighborhoods= df1['Neighborhood'].nunique()
print('Unique Postcodes : ' + str(postcodes))
print('Unique Boroughs  : '+ str(boroughs))
print('Unique Neighborhoods  :' + str(neighborhoods))

Unique Postcodes : 103
Unique Boroughs  : 11
Unique Neighborhoods  :207


In [26]:
#df1.groupby(['Postcode','Borough','Neighborhood']).size().reset_index(name='Count').head()

In [27]:
df1.head()

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M6A        North York  Lawrence Heights
4      M6A        North York    Lawrence Manor

In [28]:
nrows1=len(df1)
nrows1

210

In [29]:
#let's keep df1 intact and work with df2 as precaution
df2=df1
df2.head()

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M6A        North York  Lawrence Heights
4      M6A        North York    Lawrence Manor

In [30]:
# I will use two pointers, 'n and m' to sweep through the dataframe. So 'n' will vary from
#  0 to 211, and 'm ' will vary from 1 to 212
nrows2=len(df2)-1
nrows2

209

<h2>Loop Algorithm to extract the consolidate dataframe df2 with unique postalcodes.

In [31]:
n=0

while n < nrows2 :
    post1=df2.iloc[n,0]
    #post1
    m=n+1
    post2=df2.iloc[m,0]
    #post2
    neigh1=df2.iloc[n,2]
    neigh2=df2.iloc[m,2]
    if post1==post2:
        df2.Neighborhood[n,2] = neigh1=neigh1+','+neigh2
        #df2 = df2[df2.Neighbourhood != 'neigh2']
        df2=df2.drop(df2.index[m])
        nrows2=nrows2-1
        df2 = df2.reset_index(drop=True)
    else:
        n=n+1


#df2 = df2.reset_index(drop=True)
# print(post1, post2, 'Nigh1 is...'+ neigh1,',,,,,,,','Nigh2 is...' + neigh2, n,m)
df2.index

RangeIndex(start=0, stop=103, step=1)

In [32]:
df2.head()

Postcode           Borough  \
0      M3A        North York   
1      M4A        North York   
2      M5A  Downtown Toronto   
3      M6A        North York   
4      M7A  Downtown Toronto   

                                        Neighborhood  
0                                          Parkwoods  
1                                   Victoria Village  
2  Kingsway Park South West,Mimico NW,The Queensw...  
3                    Lawrence Heights,Lawrence Manor  
4                                       Queen's Park

<h2> DataFrame consolidates to Postalcodes sorted by postalcodes

In [33]:
df2.sort_values('Postcode')
df2.head(15)

Postcode           Borough  \
0       M3A        North York   
1       M4A        North York   
2       M5A  Downtown Toronto   
3       M6A        North York   
4       M7A  Downtown Toronto   
5       M9A      Queen's Park   
6       M1B       Scarborough   
7       M3B        North York   
8       M4B         East York   
9       M5B  Downtown Toronto   
10      M6B        North York   
11      M9B         Etobicoke   
12      M1C       Scarborough   
13      M3C        North York   
14      M4C         East York   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2   Kingsway Park South West,Mimico NW,The Queensw...  
3                     Lawrence Heights,Lawrence Manor  
4                                        Queen's Park  
5                                        Queen's Park  
6                                       Rouge,Malvern  
7                                     Don Mills North  
8                      Woodbine Gardens,Parkview Hill  
9                             Ryerson,Garden District  
10                                          Glencairn  
11  Cloverdale,Islington,Martin Grove,Princess Gar...  
12               Highland Creek,Rouge Hill,Port Union  
13                    Flemingdon Park,Don Mills South  
14                                   Woodbine Heights

In [34]:
df2.shape

(103, 3)

<h1>End of Module1

<h2>---------*---------*-----------*--------*----------*-------------*---------------*----------*----------*--------*--------*-------*------*------

<h1>Clustering Neighborhood in Toronto-2

<h2>Module2 : Now let's collect Longitude and latitude coordinates for every postal code and add them in the dataframe

In [2]:
import json
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim #converting address into longitude and latitude values

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ca-certificates-2019.11.28 |       hecc5488_0         145 KB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    certifi-2019.11.28         |           py36_0         149 KB  conda-forge
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1.21.0-py_0       conda-forge

The following packages will be UPDATED:

    ca-

In [35]:
#convert the csv into the dataframe
url='http://cocl.us/Geospatial_data'
df_pcodes=pd.read_csv(url)
df_pcodes.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [42]:
#need to change the column  name Postal Code to PostalCode for proper merger
df_pcodes.columns = ['Postcode', 'Latitude', 'Longitude']
df_pcodes.head()

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

<h2> Now let's sort df2 dataframe by postal codes and then we will merge it with df_pcodes based on common columns of postal codes

In [39]:
df2s=df2.sort_values('Postcode')
df2s.head()

Postcode      Borough                          Neighborhood
6       M1B  Scarborough                         Rouge,Malvern
12      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
18      M1E  Scarborough       Guildwood,Morningside,West Hill
22      M1G  Scarborough                                Woburn
26      M1H  Scarborough                             Cedarbrae

<h2> Table is Sorted by Postal Code

In [43]:
neighborhoods=pd.merge(df2s,df_pcodes, how='right', on = 'Postcode')
neighborhoods.head(15)

Postcode      Borough                                       Neighborhood  \
0       M1B  Scarborough                                      Rouge,Malvern   
1       M1C  Scarborough               Highland Creek,Rouge Hill,Port Union   
2       M1E  Scarborough                    Guildwood,Morningside,West Hill   
3       M1G  Scarborough                                             Woburn   
4       M1H  Scarborough                                          Cedarbrae   
5       M1J  Scarborough                                Scarborough Village   
6       M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park   
7       M1L  Scarborough                      Clairlea,Golden Mile,Oakridge   
8       M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West   
9       M1N  Scarborough                         Birch Cliff,Cliffside West   
10      M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...   
11      M1R  Scarborough                                   Maryvale,Wexford   
12      M1S  Scarborough                                          Agincourt   
13      M1T  Scarborough              Clarks Corners,Sullivan,Tam O'Shanter   
14      M1V  Scarborough  Agincourt North,L'Amoreaux East,Milliken,Steel...   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577

<h2>Table dataframe below contains 103 rows by the 5 columns requested.

In [44]:
neighborhoods.shape

(103, 5)

<h1> End of Module2

<h2>-------*--------*--------*------*-------*---------*----------*-------*-------*----------*--------*--------*----------*---------*--------

<h1> Clustering Neighborhoods in Toronto-3

<h2>Module3 : Explore and Cluster the Neighborhoods in Toronto

In [46]:

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-4.0.1               |             py_0         575 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    branca-0.4.0               |             py_0          26 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         673 KB

The following NEW packages will be INSTALLED:

    altair:  4.0.1-py_0 conda-forge
    branca:  0.4.0-py_0 conda-forge
    folium:  0.5.0-py_0 conda-forge
    vincent: 0.4.4-py_1 conda-forge


folium-0.5.0         | 45 KB     | ##################################### | 1

In [47]:
#focus on subset of all Borough that contain the word Toronto
toronto_data= neighborhoods[neighborhoods['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True) 
toronto_data.head()

Postcode          Borough                   Neighborhood   Latitude  \
0      M4E     East Toronto                    The Beaches  43.676357   
1      M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2      M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3      M4M     East Toronto                Studio District  43.659526   
4      M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [48]:
toronto_data.shape

(39, 5)

<h2> Longitude and Latitude of Toronto is:43.6532° N, 79.3832° W

In [49]:
latitude = 43.6532
longitude= -79.3832

<h2>Let's create a map of Toronto  with Borough's That contain the word Toronto

In [51]:
# create map of TORONTO using latitude and longitude values above:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<h2> Now we are going to use Foursquare API to explore the Neighborhoods and segment them

In [52]:
CLIENT_ID='YCOHHSZR403IE20DBEFKXZFEVVYC4FGBWWA2O1LLJELS1W3A'
CLIENT_SECRET='5BHDXFH3QJ4UENLFJZCBVEWVQMXJPSWUIAL3Z2PJUBP1LC22'
VERSION='20200223'

In [53]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

<H2>Let's create a function to repeat the same process to all the neighborhoods in Manhattan Lab example

In [54]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

<H2>Now write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues

In [56]:

toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Kingsway Park South West,Mimico NW,The Queensway West,Royal York South West,South of Bloor
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place

In [57]:
print(toronto_venues.shape)
toronto_venues.head()

(1720, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1                 Glen Stewart Ravine       43.676300       -79.294784   
2  The Big Carrot Natural Food Market       43.678879       -79.297734   
3                 Grover Pub and Grub       43.679181       -79.297215   
4                       Upper Beaches       43.680563       -79.292869   

         Venue Category  
0                 Trail  
1  Other Great Outdoors  
2     Health Food Store  
3                   Pub  
4          Neighborhood

<h2>Let's check how many venues were returned for each neighborhood


In [58]:
toronto_venues.groupby('Neighborhood').count()


Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            56   
Brockton,Exhibition Place,Parkdale Village                             22   
Business Reply Mail Processing Centre 969 Eastern                      15   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     17   
Cabbagetown,St. James Town                                             49   
Central Bay Street                                                     82   
Chinatown,Grange Park,Kensington Market                                84   
Christie                                                               18   
Church and Wellesley                                                   83   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             36   
Davisville North                                                        7   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     14   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            19   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      4   
Harbord,University of Toronto                                          36   
Harbourfront East,Toronto Islands,Union Station                       100   
High Park,The Junction South                                           24   
Kingsway Park South West,Mimico NW,The Queenswa...                     46   
Lawrence Park                                                           5   
Little Portugal,Trinity                                                53   
Moore Park,Summerhill East                                              4   
North Toronto West                                                     23   
Parkdale,Roncesvalles                                                  14   
Queen's Park                                                           38   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede,Swansea                                                      39   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        96   
Studio District                                                        42   
The Annex,North Midtown,Yorkville                                      22   
The Beaches                                                             5   
The Beaches West,India Bazaar                                          20   
The Danforth West,Riverdale                                            41   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             56   
Brockton,Exhibition Place,Parkdale Village                              22   
Business Reply Mail Processing Centre 969 Eastern                       15   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      17   
Cabbagetown,St. James Town                                              49   
Central Bay Street                                                      82   
Chinatown,Grange Park,Kensington Market                                 84   
Christie                                                                18   
Church and Wellesley                

<h2>Let's find out how many unique categories can be curated from all the returned venues

In [59]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))


There are 240 uniques categories.


<h2>Analyze Each Neighborhood

In [60]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols=list(toronto_onehot.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehot=toronto_onehot[['Neighborhood']+cols]

# rename Neighborhood for Neighbourhood so that future merge works
toronto_onehot.rename(columns = {'Neighborhood': 'Neighbourhood'}, inplace = True)
toronto_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0   The Beaches                  0        0                   0             0   
1   The Beaches                  0        0                   0             0   
2   The Beaches                  0        0                   0             0   
3   The Beaches                  0        0                   0             0   
4   The Beaches                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Bus Stop  Butcher  Café  Cajun / Creole Restaurant  Camera Store  \
0         0        0     0                          0             0   
1         0        0     0                          0             0   
2         0        0     0                          0             0   
3         0        0     0

In [61]:
toronto_onehot.shape


(1720, 240)

In [62]:
#let's group rows by neighborhoods by taking the mean of the frequency of occurrence of each category
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                              Adelaide,King,Richmond           0.000000   
1                                         Berczy Park           0.000000   
2          Brockton,Exhibition Place,Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
5                          Cabbagetown,St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7             Chinatown,Grange Park,Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.012048   
10                      Commerce Court,Victoria Hotel           0.000000   
11                                         Davisville           0.000000   
12                                   Davisville North           0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
14            Design Exchange,Toronto Dominion Centre           0.000000   
15                        Dovercourt Village,Dufferin           0.000000   
16              First Canadian Place,Underground city           0.000000   
17                 Forest Hill North,Forest Hill West           0.000000   
18                      Harbord,University of Toronto           0.000000   
19    Harbourfront East,Toronto Islands,Union Station           0.000000   
20                       High Park,The Junction South           0.000000   
21  Kingsway Park South West,Mimico NW,The Queensw...           0.000000   
22                                      Lawrence Park           0.000000   
23                            Little Portugal,Trinity           0.000000   
24                         Moore Park,Summerhill East           0.000000   
25                                 North Toronto West           0.000000   
26                              Parkdale,Roncesvalles           0.000000   
27                                       Queen's Park           0.000000   
28                                           Rosedale           0.000000   
29                                           Roselawn           0.000000   
30                                  Runnymede,Swansea           0.000000   
31                            Ryerson,Garden District           0.000000   
32                                     St. James Town           0.000000   
33                    Stn A PO Boxes 25 The Esplanade           0.000000   
34                                    Studio District           0.000000   
35                  The Annex,North Midtown,Yorkville           0.000000   
36                                        The Beaches           0.000000   
37                      The Beaches West,India Bazaar           0.000000   
38                        The Danforth West,Riverdale           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000000            0.000000      0.000000        0.000000   
2   0.000000            0.000000      0.000000        0.000000   
3   0.000000            0.000000      0.000000        0.000000   
4   0.058824            0.058824      0.058824        0.117647   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.000

In [63]:
toronto_grouped.shape #let's confirm the new size


(39, 240)

<h2>Let's print each neighborhood along with the top 5 most common venues


In [64]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.06
1  Thai Restaurant  0.04
2             Café  0.04
3              Bar  0.04
4   Cosmetics Shop  0.03


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1            Beer Bar  0.04
2      Farmers Market  0.04
3   French Restaurant  0.04
4  Seafood Restaurant  0.04


----Brockton,Exhibition Place,Parkdale Village----
                venue  freq
0                Café  0.14
1      Breakfast Spot  0.09
2         Coffee Shop  0.09
3  Italian Restaurant  0.05
4             Stadium  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0                 Spa  0.07
1              Garden  0.07
2  Light Rail Station  0.07
3      Farmers Market  0.07
4       Auto Workshop  0.07


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0   Airport Service  0.18
1    Airport Loun

<h2>Let's put that into a pandas dataframe
<h2>First, let's write a function to sort the venues in descending order.

In [65]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

<h2>Now let's create the new dataframe and display the top 10 venues for each neighborhood.
    

In [67]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                              Adelaide,King,Richmond           Coffee Shop   
1                                         Berczy Park           Coffee Shop   
2          Brockton,Exhibition Place,Parkdale Village                  Café   
3   Business Reply Mail Processing Centre 969 Eastern                   Spa   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   
5                          Cabbagetown,St. James Town                  Park   
6                                  Central Bay Street           Coffee Shop   
7             Chinatown,Grange Park,Kensington Market                   Bar   
8                                            Christie         Grocery Store   
9                                Church and Wellesley           Coffee Shop   
10                      Commerce Court,Victoria Hotel           Coffee Shop   
11                                         Davisville          Dessert Shop   
12                                   Davisville North                   Gym   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...                   Pub   
14            Design Exchange,Toronto Dominion Centre           Coffee Shop   
15                        Dovercourt Village,Dufferin                Bakery   
16              First Canadian Place,Underground city           Coffee Shop   
17                 Forest Hill North,Forest Hill West    Mexican Restaurant   
18                      Harbord,University of Toronto                  Café   
19    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
20                       High Park,The Junction South                  Café   
21  Kingsway Park South West,Mimico NW,The Queensw...           Coffee Shop   
22                                      Lawrence Park    Photography Studio   
23                            Little Portugal,Trinity                   Bar   
24                         Moore Park,Summerhill East            Playground   
25                                 North Toronto West        Clothing Store   
26                              Parkdale,Roncesvalles             Gift Shop   
27                                       Queen's Park           Coffee Shop   
28                                           Rosedale                  Park   
29                                           Roselawn                  Pool   
30                                  Runnymede,Swansea           Coffee Shop   
31                            Ryerson,Garden District           Coffee Shop   
32                                     St. James Town           Coffee Shop   
33                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
34                                    Studio District                  Café   
35                  The Annex,North Midtown,Yorkville        Sandwich Place   
36                                        The Beaches                   Pub   
37                      The Beaches West,India Bazaar           Pizza Place   
38                        The Danforth West,Riverdale      Greek Restaurant   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0              Thai Restaurant                  Café                   Bar   
1                     Beer Bar     French Restaurant                  Café   
2                  Coffee Shop        Breakfast Spot         Grocery Store   
3                Auto Workshop                  Park           Pizza Place   
4               Airport Lounge      Airport Terminal       Harbor / Marina   
5                         Café            Restaurant           Coffee Shop   
6           Italian Restaurant          Burger Joint        Ice Cream Shop   
7        Vietnamese Restaurant                  Café    Chinese Restaurant   
8                         Café                  Park    Italian Restaurant   
9          Japanese Restaurant      Sushi Restaurant               Gay Bar   
10                   

<h2>Cluster Neighbourhoods

<h2>Run k-means to cluster the neighbourhood into 5 clusters

In [68]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [71]:
toronto_merged = toronto_data

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postcode          Borough                   Neighborhood   Latitude  \
0      M4E     East Toronto                    The Beaches  43.676357   
1      M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2      M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3      M4M     East Toronto                Studio District  43.659526   
4      M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue  \
0 -79.293031               1                   Pub   
1 -79.352188               1      Greek Restaurant   
2 -79.315572               1           Pizza Place   
3 -79.340923               1                  Café   
4 -79.388790               1    Photography Studio   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                       Trail     Health Food Store  Other Great Outdoors   
1                 Coffee Shop    Italian Restaurant        Ice Cream Shop   
2              Sandwich Place                   Pub          Burger Joint   
3                 Coffee Shop                Bakery               Brewery   
4  Construction & Landscaping                  Park           Swim School   

  5th Most Common Venue   6th Most Common Venue 7th Most Common Venue  \
0                 Diner          Discount Store               Dog Run   
1             Bookstore  Furniture / Home Store    Frozen Yogurt Shop   
2         Burrito Place            Liquor Store                  Park   
3             Gastropub     American Restaurant    Italian Restaurant   
4              Bus Line                   Diner    Falafel Restaurant   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0         Doner Restaurant            Donut Shop            Yoga Studio  
1  Fruit & Vegetable Store             Juice Bar           Liquor Store  
2        Fish & Chips Shop            Steakhouse     Italian Restaurant  
3              Yoga Studio                 Diner                    Bar  
4              Event Space  Ethiopian Restaurant      Electronics Store

<h2>Final resulting clusters visualization - Toronto Neighborhoods

In [73]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<h1> End of Module3